api/app/signals/apps/users/v1/serializers/users.py

```diff
@@ -0,0 +1,107 @@
+from django.core.validators import EmailValidator
+from rest_framework import serializers
+
+from signals.apps.api.generics.mixins import WriteOnceMixin
+from signals.apps.users.models import Profile
+from signals.apps.users.v1.serializers.permissions import PermissionSerializer
+from signals.apps.users.v1.serializers.profiles import ProfileDetailSerializer, ProfileSerializer
+from signals.apps.users.v1.serializers.roles import RoleSerializer
+
+
+def _get_groups_queryset():
+    return Group.objects.all()
+
+
+def _get_permissions_queryset():
+    return Permission.objects.all()
+
+
+class UserListHALSerializer(WriteOnceMixin, HALSerializer):
+    _display = DisplayField()
+    roles = serializers.SerializerMethodField()
+    role_ids = serializers.PrimaryKeyRelatedField(
+        many=True, required=False, read_only=False, write_only=True,
+        queryset=_get_groups_queryset(), source='groups'
+    )
+    profile = ProfileSerializer(read_only=True)
+
+    class Meta:
+        model = User
+        fields = (
+            '_links',
+            '_display',
+            'id',
+            'username',
+            'is_active',
+            'roles',
+            'role_ids',
+            'profile',
+        )
+        extra_kwargs = {
+            'email': {'read_only': True},
+            'username': {'validators': [EmailValidator()]},
+        }
+        write_once_fields = (
+            'username',
+        )
+
+    def get_roles(self, obj):
+        return [group.name for group in obj.groups.all()]
+
+
+class UserDetailHALSerializer(WriteOnceMixin, HALSerializer):
+    _display = DisplayField()
+    roles = RoleSerializer(source='groups', many=True, read_only=True)
+    role_ids = serializers.PrimaryKeyRelatedField(
+        many=True, required=False, read_only=False, write_only=True,
+        queryset=_get_groups_queryset(), source='groups'
+    )
+    permissions = PermissionSerializer(source='user_permissions', many=True, read_only=True)
+    permission_ids = serializers.PrimaryKeyRelatedField(
+        many=True, required=False, read_only=False, write_only=True,
+        queryset=_get_permissions_queryset(), source='user_permissions'
+    )
+    profile = ProfileDetailSerializer(read_only=True)
+
+    class Meta:
+        model = User
+        fields = (
+            '_links',
+            '_display',
+            'id',
+            'username',
+            'email',
+            'first_name',
+            'last_name',
+            'is_active',
+            'is_staff',
+            'is_superuser',
+            'roles',
+            'role_ids',
+            'permissions',
+            'permission_ids',
+            'profile',
+        )
+        extra_kwargs = {
+            'email': {'read_only': True},
+            'is_staff': {'read_only': True},
+            'is_superuser': {'read_only': True},
+            'username': {'validators': [EmailValidator()]},
+        }
+        write_once_fields = (
+            'username',
```

Comment:
+ 'email',` ?

Reply:
Nope, we do not allow writing the email. It is populated with the username in the create function on the serializer. This way we ensure the username's are an email (see the validators) that we can use as the email on a user object.